R/help.R

```diff
@@ -238,33 +248,63 @@ help_formals_handler.python.builtin.object <- function(topic, source) {
   NULL
 }
 
+sphinx_doc_params_pattern <- function() { "\nParameters\n+[-]+" }
+
+sphinx_doc_params_matches <- function(doc) {
+  regexpr(pattern = sphinx_doc_params_pattern(), doc)
+}
+
+is_sphinx_doc <- function(doc) {
+  sphinx_doc_params_matches(doc)[[1]] != -1
+}
+
+sphinx_doctree_from_doc <- function(doc) {
+  module_name <- "docutils"
+  if (py_module_available(module_name)) {
+    docutils <- import(module_name)
+    docutils$core$publish_doctree(doc)
+  } else {
+    stop(paste0(module_name, " module needs to be installed to extract doc from Sphinx style documentation."))
+  }
+}
+
 # Extract argument descriptions from python docstring
 arg_descriptions_from_doc <- function(args, doc) {
-  
-  # extract arguments section of the doc and break into lines
-  arguments <- section_from_doc('Arg(s|uments)', doc)
-  doc <- strsplit(doc, "\n", fixed = TRUE)[[1]]
-  
-  arg_descriptions <- sapply(args, function(arg) {
-    arg_line <- which(grepl(paste0("^\\s+", arg, ":"), doc))
-    if (length(arg_line) > 0) {
-      line <- doc[[arg_line]]
-      arg_description <- substring(line, regexpr(':', line)[[1]] + 1)
-      next_line <- arg_line + 1
-      while((arg_line + 1) <= length(doc)) {
-        line <- doc[[arg_line + 1]]
-        if (!grepl("^\\s*$", line) && !grepl("^\\s+\\w+: ", line)) {
-          arg_description <- paste(arg_description, line)
-          arg_line <- arg_line + 1
+  if (is.null(doc)) {
+    arg_descriptions <- args
+  } else if (is_sphinx_doc(doc)) {
+    doctree <- sphinx_doctree_from_doc(doc)
+    params <- doctree$ids$parameters$children[[2]]$children
```

Comment:
Could this ever fail? (I think this is overall just an argument for us to wrap everything in `tryCatch()` to catch any unanticipated errors while calling Python code)

Reply:
I believe these are relatively stable since Sphinx cannot break backward compatibility like that. But we definitely need try catch to wrap everything.